src/shape_transform_descriptor.cpp

```diff
@@ -0,0 +1,835 @@
+    }
+    return result;
+}
+
+static std::vector<std::size_t> compute_dims(const operation& op,
+                                             const std::vector<std::size_t>& idims)
+{
+    shape s{shape::float_type, idims};
+    return op.compute_shape({s}).lens();
+}
+
+static std::vector<std::size_t> compute_dims(const std::vector<operation>& ops,
+                                             const std::vector<std::size_t>& idims)
+{
+    shape s{shape::float_type, idims};
+    for(const auto& op : ops)
+        s = op.compute_shape({s});
+    return s.lens();
+}
+
+bool shape_transform_descriptor::apply(const std::vector<operation>& ops)
+{
+    std::vector<std::size_t> dims;
+    std::transform(dimensions.begin(),
+                   dimensions.end(),
+                   std::back_inserter(dims),
+                   [](const dimension& d) { return d.len(); });
+    for(const auto& op : ops)
+    {
+        auto v = op.to_value();
+        if(contains({"reshape", "squeeze", "unsqueeze", "flatten"}, op.name()))
+        {
+            dims = compute_dims(op, dims);
+            if(not apply_reshape(dims))
+                return false;
+        }
+        else if(op.name() == "transpose")
+        {
+            dims = compute_dims(op, dims);
+            if(not apply_transpose(v["permutation"].to_vector<std::int64_t>()))
+                return false;
+        }
+        else if(op.name() == "multibroadcast")
+        {
+            dims = compute_dims(op, dims);
+            // cppcheck-suppress knownConditionTrueFalse
+            if(not apply_broadcast(dims))
+                return false;
+        }
+        else if(op.name() == "broadcast")
+        {
+            dims = compute_dims(op, dims);
+            // cppcheck-suppress knownConditionTrueFalse
+            if(not apply_broadcast(dims, v["axis"].to<std::size_t>()))
+                return false;
+        }
+        else if(op.name() != "contiguous")
+        {
+            return false;
+        }
+    }
+    return true;
+}
+bool shape_transform_descriptor::apply_reshape(const std::vector<std::size_t>& rdims)
+{
+    assert(migraphx::elements(rdims) == this->elements());
+    std::vector<dimension> new_dims;
+    auto subs     = get_all_subdimensions(dimensions);
+    std::size_t i = 0;
+    std::size_t r = 0;
+    while(i < subs.size() and r < rdims.size())
+    {
+        const auto& sub = subs[i];
+        auto idim       = sub.len;
+        auto rdim       = rdims[r];
+        if(idim == rdim)
+        {
+            new_dims.push_back({{sub}});
+        }
+        // squeeze
+        else if(rdim > idim)
+        {
+            auto start = subs.begin() + i;
+            auto it = compute_end_dim(start, subs.end(), rdim, std::mem_fn(&dimension::sub::len));
+            if(it == start)
+                return false;
+            auto n = it - start;
+            i += n;
+            new_dims.push_back({{start, it + 1}});
```

Comment:
I don't think this is right. `it` should point to the end dimension already? Somehow it works right now. Using LLDB it seems that this code is optimized out. But when I put in a new test and debug statements I get out of range segmentation faults.

Here's the test I added: 
```
TEST_CASE(record_reshape_merge)
{
    auto desc = make_descriptor({3, 4, 5}, make_op("reshape", {{"dims", {3, 20}}}));
    EXPECT(get_final_lens(desc) == final_lens{3, 20});
    EXPECT(get_all_lens(desc) == all_lens{{3}, {4, 5}});
    EXPECT(get_all_axes(desc) ==
           all_axes{d_axes{{0}}, d_axes{{1}, {2}}});
}
```

I added these debug statements:
```
            auto n = it - start;
            std::cout << "n: " << n << "\n";
            std::cout << "i before: " << i << "\n";
            i += n;
            std::cout << "i after: " << i << "\n";
            std::cout << "it + 1 - start: " << it + 1 - start << "\n";
            std::cout << "(it + 1)->len" << (it + 1)->len << "\n";
            std::cout << "(it + 1)->axis.front()" << (it + 1)->axis.front() << "\n";
            new_dims.push_back({{start, it + 1}});
```

Reply:
Figured it out, using begin to end iterators